thrust/thrust/iterator/offset_iterator.h

```diff
@@ -0,0 +1,157 @@
+//! dereferencing, comparing, or computing the distance between two offset_iterators. This is useful, when the
+//! underlying iterator cannot be incremented, decremented, or advanced (e.g., because those operations are only
+//! supported in device code).
+//!
+//! The following code snippet demonstrates how to create a \p offset_iterator:
+//!
+//! \code
+//! #include <thrust/iterator/offset_iterator.h>
+//! #include <thrust/fill.h>
+//! #include <thrust/device_vector.h>
+//!
+//! int main()
+//! {
+//!   thrust::device_vector<int> data{1, 2, 3, 4};
+//!   auto b = offset_iterator{data.begin(), 1};
+//!   auto e = offset_iterator{data.end(), -1};
+//!   thrust::fill(b, e, 42);
+//!   // data is now [1, 42, 42, 4]
+//!   ++b; // does not call ++ on the underlying iterator
+//!   assert(b == e - 1);
+//!
+//!   return 0;
+//! }
+//! \endcode
+//!
+//! Combining a \p offset_iterator with \p cub::FutureValue can also be used to retrieve the offset from an iterator.
+//! However, such an \p offset_iterator cannot be moved anymore, since changes to the offset can not be written back.
+//!
+//! \code
+//! #include <thrust/iterator/offset_iterator.h>
+//! #include <thrust/fill.h>
+//! #include <thrust/functional.h>
+//! #include <thrust/device_vector.h>
+//! #include <cub/util_type.h>
+//!
+//! int main()
+//! {
+//!   using thrust::placeholders::_1;
+//!   thrust::device_vector<int> data{1, 2, 3, 4};
+//!
+//!   thrust::device_vector<ptrdiff> offsets{1}; // offset is only available on device
+//!   auto offset = cub::FutureValue{thrust::make_transform_iterator(offsets.begin(), _1 * 2)};
+//!   thrust::offset_iterator iter(v.begin(), offset); // load and transform offset upon access
+//!   // iter is at position 2 (= 1 * 2) in data, and would return 3 in device code
+//!
+//!   return 0;
+//! }
+//! \endcode
+//!
+//! In the above example, the offset is loaded from a device vector, transformed by a \p transform_iterator, and then
+//! applied to the underlying iterator, when the \p offset_iterator is accessed.
+template <typename Iterator, typename Offset = typename ::cuda::std::iterator_traits<Iterator>::difference_type>
+class offset_iterator : public iterator_adaptor<offset_iterator<Iterator, Offset>, Iterator>
+{
+  //! \cond
+  friend class iterator_core_access;
+  using super_t = iterator_adaptor<offset_iterator<Iterator, Offset>, Iterator>;
+
+public:
+  using reference       = typename super_t::reference;
+  using difference_type = typename super_t::difference_type;
+  //! \endcond
+
+  _CCCL_HOST_DEVICE offset_iterator(Iterator it = {}, Offset offset = {})
+      : super_t(::cuda::std::move(it))
+      , m_offset(offset)
+  {}
+
+  _CCCL_HOST_DEVICE const Offset& offset() const
+  {
+    return m_offset;
+  }
+
+  _CCCL_HOST_DEVICE Offset& offset()
+  {
+    return m_offset;
+  }
+
+  //! \cond
+
+private:
+  _CCCL_HOST_DEVICE reference dereference() const
```

Comment:
Should this also be marked with `_CCCL_EXEC_CHECK_DISABLE` as it relies on the base iterator?